src/Output/ProgressBar.php

```diff
@@ -0,0 +1,319 @@
+     *
+     * @return void
+     * @throws UnexpectedValueException
+     */
+    public function current(int $current, string $label = '')
+    {
+        if ($this->total == 0) {
+            // Avoid dividing by 0
+            throw new UnexpectedValueException('The progress total must be greater than zero.');
+        }
+
+        if ($current > $this->total) {
+            throw new UnexpectedValueException('The current is greater than the total.');
+        }
+
+        $this->drawProgressBar($current, $label);
+
+        $this->current = $current;
+        $this->label   = $label;
+    }
+
+    /**
+     * Increments the current position we are at and re-writes the progress bar
+     *
+     * @param integer $increment The number of items to increment by
+     */
+    public function advance(int $increment = 1, string $label = '')
+    {
+        $this->current($this->current + $increment, $label);
+    }
+
+    /**
+     * Force the progress bar to redraw every time regardless of whether it has changed or not
+     */
+    public function forceRedraw(bool $force = true): self
+    {
+        $this->forceRedraw = $force;
+
+        return $this;
+    }
+
+
+    /**
+     * Update a progress bar using an iterable.
+     *
+     * @param iterable $items Array or any other iterable object
+     * @param callable $callback A handler to run on each item
+     */
+    public function each($items, callable $callback = null)
+    {
+        if ($items instanceof \Traversable) {
+            $items = iterator_to_array($items);
+        }
+
+        if (0 === $total = count($items)) {
+            return;
+        }
+
+        $this->total($total);
+
+        foreach ($items as $key => $item) {
+            if ($callback) {
+                $label = $callback($item, $key);
+            } else {
+                $label = null;
+            }
+
+            $this->advance(1, (string) $label);
+        }
+    }
+
+
+    /**
+     * Draw the progress bar, if necessary
+     */
+    protected function drawProgressBar(int $current, string $label)
+    {
+        $percentage = $this->percentageFormatted($current / $this->total);
+
+        if ($this->shouldRedraw($percentage, $label)) {
+            $progress_bar = $this->getProgressBar($current, $label);
+            $this->writer->colors($progress_bar . '<eol>');
+        }
+
+        $this->currentPercentage = $percentage;
+    }
+
+    /**
+     * Build the progress bar str and return it
+     */
+    protected function getProgressBar(int $current, string $label): string
+    {
+        if ($this->firstLine) {
+            // Drop down a line, we are about to
+            // re-write this line for the progress bar
+            $this->writer->write('');
+            $this->firstLine = false;
+        }
+
+        // Move the cursor up and clear it to the end
+        $line_count = $this->hasLabelLine ? 2 : 1;
+
+        $progress_bar  = $this->cursor->up($line_count);
+        $progress_bar .= "\r";
```

Comment:
this this CR char required? in unix it doesnt seem to be required

(i'm on branch [81-progress-bar](https://github.com/adhocore/php-cli/blob/81-progress-bar) trying to write tests)

Reply:
I haven't tried with unix but with windows it is essential. when removed, the progress bar goes to a new line on each iteration

Preview => https://ibb.co/W0kMTcz